# Loading the dataset

In [2]:
from keras.datasets import mnist

In [3]:
(train_images,train_labels),(test_images,test_labels) = mnist.load_data()

In [4]:
train_images.shape

(60000, 28, 28)

In [7]:
train_images[0][4:8]

array([[  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   3,
         18,  18,  18, 126, 136, 175,  26, 166, 255, 247, 127,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,   0,  30,  36,  94, 154, 170,
        253, 253, 253, 253, 253, 225, 172, 253, 242, 195,  64,   0,   0,
          0,   0],
       [  0,   0,   0,   0,   0,   0,   0,  49, 238, 253, 253, 253, 253,
        253, 253, 253, 253, 251,  93,  82,  82,  56,  39,   0,   0,   0,
          0,   0]], dtype=uint8)

In [8]:
train_labels[:10]

array([5, 0, 4, 1, 9, 2, 1, 3, 1, 4], dtype=uint8)

In [9]:
test_images.shape

(10000, 28, 28)

# Network

In [10]:
from keras import models, layers

In [37]:
network = models.Sequential()
network.add(layers.Dense(512, activation='relu', input_shape=(28 * 28,)))
network.add(layers.Dense(10, activation='softmax'))

The number of units is 65% less than the input_shape:

In [12]:
28*28

784

In [24]:
(28*28)/2

392.0

In [14]:
512/(28*28)

0.6530612244897959

* Dense layer is for a 'fully connected' neural layer.
* Softmax is for returning a normalized probability over the 10 classes.

In [38]:
# Compilation step
network.compile(optimizer='rmsprop',
               loss='categorical_crossentropy',
               metrics=['accuracy'])

## Data preprocessing
 * reshaping
 * scaling

In [16]:
train_images = train_images.reshape((60000, 28*28))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28*28))
test_images = test_images.astype('float32') / 255

* One-hot encoding the labels

In [17]:
from keras.utils import to_categorical

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

In [18]:
train_labels[0]

array([ 0.,  0.,  0.,  0.,  0.,  1.,  0.,  0.,  0.,  0.])

# Training

In [39]:
network.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
60000/60000 [==============================] - 4s - loss: 0.2535 - acc: 0.9263     
Epoch 2/5
60000/60000 [==============================] - 4s - loss: 0.1021 - acc: 0.9702     
Epoch 3/5
60000/60000 [==============================] - 4s - loss: 0.0674 - acc: 0.9797     
Epoch 4/5
60000/60000 [==============================] - 4s - loss: 0.0485 - acc: 0.9854     
Epoch 5/5
60000/60000 [==============================] - 4s - loss: 0.0378 - acc: 0.9889     


# Evaluation

In [40]:
test_loss, test_acc = network.evaluate(test_images, test_labels)
print('\ntest_acc:', test_acc)

 9056/10000 [==========================>...] - ETA: 0s
test_acc: 0.981


Comparing train accuracy (0.99) to the test accuracy (0.97) suggest overfitting.